datalake_api/querier.py

```diff
@@ -210,20 +210,22 @@ def query_by_time(self, start, end, what, where=None, cursor=None):
             buckets = buckets[i:]
 
         for b in buckets:
-            kwargs = self._prepare_time_bucket_kwargs(b, what,
-                                                      limit=MAX_RESULTS/2)
-            if where is not None:
-                self._add_range_key_condition(kwargs, where)
-            if cursor is not None:
-                self._add_cursor_conditions(kwargs, cursor)
-
-            response = self._table.query(**kwargs)
-            results += self._exclude_outside(response['Items'], start, end)
-            # we _could_ deduplicate the results here to make more headroom for
-            # another bucket.
-            cursor = self._cursor_for_time_query(response, results, b)
-            if cursor is not None:
-                break
+            while len(results) <= MAX_RESULTS/2:
+                kwargs = self._prepare_time_bucket_kwargs(b, what,
+                                                          limit=MAX_RESULTS/2)
+                if where is not None:
+                    self._add_range_key_condition(kwargs, where)
+                if cursor is not None:
+                    self._add_cursor_conditions(kwargs, cursor)
+
+                response = self._table.query(**kwargs)
+                results += self._exclude_outside(response['Items'], start, end)
+                # we _could_ deduplicate the results here to make more headroom
+                # for another bucket.
+                cursor = self._cursor_for_time_query(response, results, b)
+                if cursor is None:
+                    # no more results in the bucket
+                    break
```

Comment:
Let's see if we can make this more straightforward by just tuning the limit parameter properly. I'm not really sure why I had all of this `MAX_RESULTS/2` business when we have a limit parameter at our disposal. How about something like this:

```
for b in buckets:
    headroom = MAX_RESULTS - len(results)
    if headroom <= 0 or cursor is None:
        break
    kwargs = dict(
        bucket=b,
        what=what,
        start=start, 
        end=end,
        where=where,
        limit=headroom, cursor
    )
    response = self._query_time_bucket(**kwargs))
    results += response['Items']
    cursor = self._cursor_for_time_query(response, results, b)
```

_query_time_bucket would have its own inner loop that repeatedly queries the bucket if there's more data and more headroom. Unfortunately, I don't think we are going to be able to articulate the behavior of `_exclude_outside()` as a filter condition. So in the case where there are more than MAX_RESULTS of the desired `what`s in a time bucket that do not lie within the requested time window we have to do multiple queries in that bucket.